File: src/com/gnarly/engine/display/Camera.java
package com.gnarly.engine.display;

import org.joml.Matrix4f;
import org.joml.Vector3f;

public class Camera {
	
	private Matrix4f projection, projView;
	
	private float width, height;
	private Vector3f position;
	private float rotation;
	
	public Camera(float width, float height) {
		setDims(width, height);
		position = new Vector3f();
		rotation = 0;
		projView = new Matrix4f();
	}
	
	public void setDims(float width, float height) {
		this.width = width;
		this.height = height;
		projection = new Matrix4f().setOrtho(0, width, height, 0, 0, 1);
	}
	
	public void update() {
		projection.translate(position.negate(new Vector3f()), projView);
		projView.rotateZ(-rotation);
	}
	
	public Matrix4f getProjection() {
		return new Matrix4f(projection);
	}
	
	public Matrix4f getMatrix() {
		return new Matrix4f(projView);
	}
	
	public float getX() {
		return position.x;
	}
	
	public float getY() {
		return position.y;
	}
	
	public Vector3f getPosition() {
		return new Vector3f(position);
	}
	
	public float getWidth() {
		return width;
	}
	
	public float getHeight() {
		return height;
	}
	
	public void setX(float x) {
		position.x = x;
	}
	
	public void setY(float y) {
		position.y = y;
	}
	
	public void setPosition(float x, float y) {
		position.set(x, y, position.z);
	}
	
	public void setPosition(Vector3f position) {
		this.position.x = position.x;
		this.position.y = position.y;
	}
	
	public void setCenter(float x, float y) {
		position.set(x - width / 2, y - height / 2, position.z);
	}
	
	public void setCenter(Vector3f position) {
		this.position.x = position.x - width / 2;
		this.position.y = position.y - height / 2;
	}
	
	public void translate(float x, float y, float z) {
		position.add(x, y, z);
	}
	
	public void translate(Vector3f transform) {
		position.add(transform);
	}
	
	public void setRotation(float angle) {
		rotation = angle;
	}
	
	public void rotate(float angle) {
		rotation += angle;
	}
}

File: src/com/gnarly/engine/Camera.kt
package com.gnarly.engine

import org.joml.Matrix4f
import org.joml.Vector3f
import kotlin.math.absoluteValue

class Camera {
	private val projection = Matrix4f()
	private val view = Matrix4f()
	private val viewProjection = Matrix4f()
	val model = Matrix4f()
	var width = 0.0f
		private set
	var height = 0.0f
		private set

	private val position = Vector3f()
	var rotation = 0.0f
	var scale = 1.0f

	fun setDims(left: Float, right: Float, down: Float, up: Float): Camera {
		this.width = (right - left).absoluteValue
		this.height = (up - down).absoluteValue
		projection.setOrtho(left, right, down, up, 0f, 1f)
		return this
	}

	fun update() {
		view.scaling(1.0f / scale)
			.rotateZ(-rotation)
			.translate(-position.x, -position.y, 0.0f)

		projection.mul(view, viewProjection)
	}

	var x: Float
		get() = position.x
		set(x) {
			position.x = x
		}
	var y: Float
		get() = position.y
		set(y) {
			position.y = y
		}

	fun getPosition(): Vector3f {
		return Vector3f(position)
	}

	fun setPosition(x: Float, y: Float) {
		position[x, y] = position.z
	}

	fun setPosition(vector: Vector) {
		this.position.x = vector.x
		this.position.y = vector.y
	}

	fun translate(x: Float, y: Float, z: Float) {
		position.add(x, y, z)
	}

	fun translate(transform: Vector3f?) {
		position.add(transform)
	}

	fun rotate(angle: Float) {
		rotation += angle
	}

	fun projectionView(): Matrix4f {
		return viewProjection
	}

	fun projection(): Matrix4f {
		return projection
	}

	/* MODEL */

	fun model(x: Float, y: Float, width: Float, height: Float): Matrix4f {
		return model.translation(x, y, 0.0f).scale(width, height, 0f)
	}

	fun model(x: Float, y: Float, scale: Float): Matrix4f {
		return model.translation(x, y, 0.0f).scale(scale)
	}

	fun modelRotate(x: Float, y: Float, scale: Float): Matrix4f {
		return model.translation(x, y, 0.0f).scale(scale)
	}

	fun model(x: Float, y: Float, width: Float, height: Float, rotation: Float): Matrix4f {
		return model.translation(x, y, 0.0f).rotateZ(rotation).scale(width, height, 0f)
	}

	fun modelRotateCentered(x: Float, y: Float, width: Float, height: Float, rotation: Float): Matrix4f {
		return model.translation(x, y, 0.0f).scale(width, height, 0f).translate(0.5f, 0.5f, 0.0f).rotateZ(rotation).translate(-0.5f, -0.5f, 0.0f)
	}

	fun modelCentered(x: Float, y: Float, width: Float, height: Float): Matrix4f {
		return model.translation(x, y, 0.0f).scale(width, height, 0f).translate(-0.5f, -0.5f, 0.0f)
	}
}
